test/sparsity/test_marlin.py

```diff
@@ -0,0 +1,132 @@
+    def test_quant_sparse_marlin_layout_eager(self):
+        torch.manual_seed(0)
+
+        input = torch.randn((32, 16, 4096), dtype=torch.float16, device="cuda")
+        model = (
+            nn.Sequential(
+                nn.Linear(4096, 21504),
+                nn.Linear(21504, 4096),
+                nn.ReLU(),
+                nn.Linear(4096, 21504),
+                nn.Linear(21504, 4096),
+            )
+            .half()
+            .cuda()
+        )
+
+        apply_fake_sparsity(model)
+        model_copy = copy.deepcopy(model)
+
+        # Quantized
+        quantize_(model_copy.bfloat16(), int4_weight_only())
+        dense_result = model_copy(input.bfloat16()).half()
+
+        # Sparse + quantized
+        quantize_(model, int4_weight_only(layout_type=MarlinSparseLayoutType()))
+        sparse_result = model(input)
+
+        assert torch.allclose(dense_result, sparse_result, atol=3e-1), "Results are not close"
+
+    @pytest.mark.skipif(not torch.cuda.is_available(), reason="Need CUDA available")
+    def test_quant_sparse_marlin_layout_compile(self):
+        torch.manual_seed(0)
+
+        input = torch.randn((32, 16, 4096), dtype=torch.float16, device="cuda")
+        model = (
+            nn.Sequential(
+                nn.Linear(4096, 21504),
+                nn.Linear(21504, 4096),
+                nn.ReLU(),
+                nn.Linear(4096, 21504),
+                nn.Linear(21504, 4096),
+            )
+            .half()
+            .cuda()
+        )
+
+        apply_fake_sparsity(model)
+        model_copy = copy.deepcopy(model)
+
+        # Quantized
+        quantize_(model_copy.bfloat16(), int4_weight_only())
+        model_copy.foward = torch.compile(model_copy.forward, fullgraph=True)
+        dense_result = model_copy(input.bfloat16()).half()
+
+        # Sparse + quantized
+        quantize_(model, int4_weight_only(layout_type=MarlinSparseLayoutType()))
+        model.forward = torch.compile(model.forward, fullgraph=True)
+        sparse_result = model(input)
+
+        assert torch.allclose(dense_result, sparse_result, atol=3e-1), "Results are not close"
+
+    @pytest.mark.skipif(not torch.cuda.is_available(), reason="Need CUDA available")
+    def test_pack_unpack_equivalence(self):
+        num_bits = 4
+        group_size = 128
+        shape = (11008, 4096)
+        max_q_val = 2**num_bits - 1
+        half_q_val = (max_q_val + 1) // 2
+
+        w = torch.rand(shape, dtype=torch.float16, device="cuda")
+        size_k, size_n = w.shape
+
+        # Inject 2:4 sparsity mask
+        w_24, _ = inject_24(w, *w.shape)
+
+        # Quantize weights 
+        w_24 = w_24.reshape((-1, group_size, size_n))
+        w_24 = w_24.permute(1, 0, 2)
+        w_24 = w_24.reshape((group_size, -1))
+
+        # Compute scale for each group
+        scales = torch.max(torch.abs(w_24), 0, keepdim=True)[0]
+        scales *= 2 / max_q_val  # 2 => symmetric
```

Comment:
these seems to be complicated, should we add a test to make sure the quantized weight and calculate scale matches the result of using affine quantized tensor quant primitives in `int4_weight_only` with the `MarlinSparseLayoutType`?

Reply:
Yeah you are right. Just updated the test to instead use the quantization APIs from `int4_weight_only`

Reply:
if this becomes a duplicate with test_sparse_api, maybe we can remove this one I feel